modules/core/petra/petra-http-invoker/src/main/java/com/liferay/petra/http/invoker/HttpInvoker.java

```diff
@@ -210,23 +211,29 @@ protected HttpURLConnection getHttpURLConnection(
 			(HttpURLConnection)url.openConnection();
 
 		try {
-			HttpURLConnection setMethodHttpURLConnection = httpURLConnection;
+			httpURLConnection.setRequestMethod(httpMethod.name());
+		}
+		catch (ProtocolException protocolException) {
```

Comment:
@drewbrokke this is a bad practice, we should not use exception to do execution flow control. At least we know exactly which methods the setter can handle, and for others we will have to use reflection. If you want to stick to setter whenever possible, it should be done via method type checking, rather than catching this exception then do a reflection fallback.

But on a 2nd thought about using the setter, what's the point of doing it? It is not faster than reflection call, it could actually be slower due to the internal impl of setRequestMethod(). Doing so will branch our code into 2 paths which is not good from maintaining point of view. In case of future jdk internal impl change that could break the reflection call, we would want to see it on all methods, rather than just one method. This way it will raise the alarm for immediate attention.

Reply:
> we should not use exception to do execution flow control

Generally I agree about catch-as-control-flow, but If we want to try the setter first in all cases, then I don't see a different way that does not involve some sort of reflection-based (and impl-based) introspection.

> it should be done via method type checking

Would you be willing to expand on your thoughts here? This would be looking at the available method field of the connection through reflection?

> Doing so will branch our code into 2 paths which is not good from maintaining point of view

Agree, but we're already branching the reflection, so it's already "not ideal". Once we drop support for the JDK that does not support PATCH, we can safely remove the fallback.

> In case of future jdk internal impl change that could break the reflection call...

This is my primary reason for preferring the setter. It minimizes impl-specific code, so there is less chance of unexpected breakage if switching to a different JDK. The impl-specific reflection call is the cause of this problem in the first place. 

> we would want to see it on all methods, rather than just one method.

Can you expand on this a bit more? I don't think I fully understand what you mean by this.

In any case, thank you for raising these points.